src/box/lua/net_cluster.lua

```diff
@@ -0,0 +1,466 @@
+-- This module provides connections to all replicasets of the cluster.
+-- It maintains a list of connections (that actually are net.replicaset
+-- connections) and allows to subscribe to an event (aka watch) on entire
+-- cluster.
+--
+-- There two possible ways to use this module.
+-- 1. Use common shared object with default options. This object is created
+--  on demand. This module has `get`, `watch_leaders` etc methods that actually
+--  redirected to the shared object.
+-- 2. Create and use personal object via `cluster_new` module method. On this
+--  way some options can be specified. Such an object will have `close`, `get`,
+--  `watch_leaders` etc method.
+local fiber = require('fiber')
+local net_replicaset = require('internal.net.replicaset')
+
+local cluster_cfg_defaults = {
+    config_mode = 'auto',
+    connect_opts = {
+        reconnect_timeout = 1,
+    }
+}
+
+local default_cluster
+
+-- Helper for lazy connection creation.
+local function acquire_replicaset_connection(cluster, replicaset)
+    local conn = replicaset.conn
+    if conn then
+        return conn
+    end
+    conn = net_replicaset.connect(replicaset.connect_cfg or replicaset.name,
+            cluster._cfg.connect_opts)
+    replicaset.conn = conn
+    return conn
+end
+
+-- Helper that closes connection if necessary.
+local function close_replicaset_connection(replicaset)
+    if replicaset.conn then
+        replicaset.conn:close()
+        replicaset.conn = nil
+    end
+end
+
+-- Check that replicaset roles intersects with given roles.
+local function replicaset_comply_roles(replicaset, roles)
+    if not roles then
+        -- No filter.
+        return true
+    end
+    if not replicaset.roles then
+        -- No desired roles.
+        return false
+    end
+    for _, role in pairs(roles) do
+        if replicaset.roles[role] then
+            return true
+        end
+    end
+    return false
+end
```

Comment:
In `experimantal.connpool` the `roles` [filter](https://github.com/tarantool/tarantool/blob/2467e421cbfd42d0f58c0d27e4e397e730412e6c/src/box/lua/experimental/connpool.lua#L132-L150) verifies that an instance has all the provided roles (AND semantic). Here we have the opposite: a replicaset should have any of the provided roles (OR semantic).

I don't know for sure what is better, but I would like to make the same named filter working consistently.

Personally, I think that the AND semantics is more intuitive here: look up an instance/a replicaset that match all the given criteria.